File: traits/traits.rs
// Implementation of trait examples from README.md

// Define the Speak trait
trait Speak {
    fn speak(&self, msg: &str);
    fn get_age(&self) -> u32;
}

// Define the Dog struct
struct Dog {
    age: u32,
}

// Implement Speak trait for Dog
impl Speak for Dog {
    fn speak(&self, msg: &str) {
        println!("{}", msg);
    }

    fn get_age(&self) -> u32 {
        self.age
    }
}

fn main() {
    // Create a Dog instance
    let dog = Dog { age: 5 };

    // Use the trait methods
    dog.speak("Woof!");
    println!("Dog's age is: {}", dog.get_age());

    // Demonstrate using the trait in a generic function
    fn make_speak<T: Speak>(speaker: &T, message: &str) {
        speaker.speak(message);
        println!("Speaker's age: {}", speaker.get_age());
    }

    make_speak(&dog, "Generic woof!");
}

File: traits/traits.cpp
#include <iostream>
#include <string_view>
#include <concepts>
#include <cassert>

// Define the Speak concept (equivalent to Rust's trait)
template <typename T>
concept Speak = requires(const T &x) {
    { x.speak(std::string_view{}) } -> std::same_as<void>;
    { x.get_age() } -> std::same_as<uint32_t>;
};

// Define the Dog struct with inline implementation
struct Dog
{
    uint32_t age;

    void speak(std::string_view msg) const
    {
        std::cout << msg << '\n';
    }

    uint32_t get_age() const
    {
        return age;
    }
};

// Verify Dog satisfies Speak concept at compile time
static_assert(Speak<Dog>);

// Generic function that works with any type that satisfies Speak
template <Speak T>
void make_speak(const T &speaker, std::string_view message)
{
    speaker.speak(message);
    std::cout << "Speaker's age: " << speaker.get_age() << std::endl;
}

int main()
{
    // Create a Dog instance
    Dog dog{5};

    // Use the methods directly
    dog.speak("Woof!");
    std::cout << "Dog's age is: " << dog.get_age() << std::endl;

    // Use the generic function
    make_speak(dog, "Generic woof!");

    return 0;
}
